Add subgraph-related functions.

scolobb opened the pull request:
This pull request adds `gfiltermap` which allows building a graph out of certain transformed contexts of the original one.  It then adds `nfilter'` and `nfilter` which allow constructing a subgraph by filtering by node labels or by node numbers.  Finally, the function `subgraph` proper is added.

ivan-m commented:
Please convert the tests into QuickCheck-based (or HSpec-based if necessary) and add them to the appropriate modules.

I think the "Add a test for *foo*" commits can be squashed into the general "Add *foo*" commits.

scolobb commented:
Here's my update.

Given that the added functions are all small, I squashed everything into two commits: one adding the functions, and the other one adding the tests.  I can surely unsquash the commits, if you prefer.

I renamed the filter functions: `nfilter` filters by `Node`s, `labnfilter` by `LNode`s, `labfilter` by node labels.  Even though `nfilter` can indeed be seen as a special case of `labnfilter`, I felt that implementing `labnfilter` before all other filters would make things slightly more complicated than if I implemented it in terms of `nfilter`, so I kind of kept the old scheme partially: `nfilter` is based on `gfiltermap`, `labnfilter` is based on `nfilter`, and `labfilter` is a special case of `labfilter`.

ivan-m commented:
I think it makes more sense (and more efficient) for `nfilter` to depend upon `labnfilter`, though I can see where the problem is in terms of how to deal with filtering the adjacency list.

I think I originally asked for them (possibly in a comment in one of the original commits? can't see it now though), but I think that after thinking about it some more `labfilter` and `labnfilter` aren't really worth it.

Then squash the commits together into one, and just call it something like "Add graph filtering functions".

scolobb commented:
> Please start the contents of where clauses on the next line indented by 2 spaces.

> Rather than providing the vs as a parameter, make the property be in Gen Bool and obtain a sub-list of nodes with sublistOf (nodes g).

Done.

Yeah, you did suggest implementing `nfilter` in terms of `labnfilter` and you did point out that, to implement `labnfilter`, I would have to use `delNodes` to remove the nodes which do not pass the filter, because one could not filter lists of `Adj b` directly.  (The comments got removed because of the rebase.)  I thought that, instead of using `delNodes`, I would prefer implementing `labnfilter` using `nfilter` and keeping a `Map` of node labels.  Do you still prefer having it the other way round (i.e., defining `nfilter` as a special case of `labnfilter`)?

> but I think that after thinking about it some more `labfilter` and `labnfilter` aren't really worth it.

Do you mean that we want to ditch the two?

> Then squash the commits together into one, and just call it something like "Add graph filtering functions".

Do you want me to both add the filtering functions and the tests for them in a single commit?

ivan-m commented:
I've been thinking about this the past few days.

My thinking is that - unless we can directly filter on the `LNode` (rather than having to keep fetching the label) it isn't worth it.  The only feasible approach to do this would be to do something like `flip delNodes g . filter (not . p) $ labNodes g`; whether this is a good approach to take is a different story.

Doing it this way means that it requires two passes through the graph (once to get the nodes, then to delete the ones that don't match).  In fact, it could be even worse due to how edge deletion currently works.

That said, the approach you've taken of using `ufold` isn't much better (it only requires one explicit pass, but the edge deletion bit is the same because of how `match` works).  I'm also unsure of `gfiltermap` being exported because of how it's possible to break the graph by adding references to non-existent nodes (then again, this is something inherent in `&` so it's not a huge problem per-se).

What do you think; is defining `labnfilter` by deleting nodes that doesn't match worth it?  The `O(..)` complexity is the same.

---

With regards to the other aspects:  I think this can be reduced to a single commit with a message about adding functions to induce subgraphs either directly or by filtering on nodes (rather than listing each function directly); add some more details in the longer message.

scolobb commented:
Oh, I wasn't aware that the `ufold`-based approached had the same asymptotic complexity as the `delNodes`-based one.  In this case I agree that implementing `labnfilter` using your idea is better, since it allows defining `nfilter` and `labfilter` as straightforward special cases, and is also simpler.  I'll do that right now.

I'd keep `gfiltermap` exported for now since it seems to be a pretty generic tool and it does not open more opportunities to shoot oneself in the foot than `&` does.  If you do lean towards seeing `gfiltermap` on the more dangerous side however, I will not oppose to hiding it.

I will also squash the commits into one and add a longer commit message.

ivan-m commented:
Well, I'm pretty sure it has the same complexity...

First of all, let's ignore implementation, or else we'd need to add `log` factors in everywhere (plus if we get a viable array/vector-based approach it might not have those factors).

For every node, it has to examine every adjacent edge to get the `Context`, then do so again for the `&`.  So it's `O(|V| + |E|)` (the `|E|` is because each edge will be examine at least once to put it into a `Context`).  If you keep every node, then each node will be `match`d and then put back in; if you delete every node then you only need to `match` each one.

The delete-based approach will only examines the edges when doing the `match` for `delNode` and not for the `&`, but it still needs to get the nodes (`O(|V|)`), filter them all (`O(|V|)`) and then delete what's left (potentially `O(|V|)` if we're deleting them all).  Deletion will require less examination of the edges.  However, overall it's still `O(|V| + |E|)` (if we _do_ take into account the log factors, then this has less of them as we only have the log factors for the nodes for the ones we're deleting; we also don't need to filter the adjacency lists).

So I think the `delNode`-based approach isn't any worse and potentially better.

Agreed on keeping `gfiltermap` though.

scolobb commented:
Oh, thank you for the detailed analysis!  I was definitely forgetting that `&` is more complex than cons `:`, and was pretty sure that `ufold` would require less pattern matching effort than deleting nodes, but now I see that I just hadn't taken my time to think it over.  Indeed, using `delNodes` may actually outperform my present version.

The new implementation should be online in less than an hour (I hope).

scolobb commented:
Here we go.

scolobb commented:
OK, I need to disappear for about two hours.  Sorry for making a mess with this series of commits.

ivan-m commented:
That's fine :)

Can you split up the renaming of the condendastion-based tests into a separate pull request (or even a separate commit)?

(I'm also going to make a prediction that it will fail to build because you're renamed them in the `TestSuite` file but not in the `Properties` file...)

scolobb commented:
Oh, okay, never do things in a hurry, that's an old piece of wisdom I'll never learn to use :-) I'll fix this pull request now and will make a separate one fixing the tests related to condensation.

scolobb commented:
All right, I believe _this_ latest commit botches nothing :-)

scolobb commented:
Sorry, my mouse slipped :-(

ivan-m commented:
Ta :)

(I indeed couldn't find anything botch-y in this commit :p)
